splice_encoding/src/main/java/com/splicemachine/encoding/BigDecimalEncoding.java

```diff
@@ -185,71 +185,51 @@ public static BigDecimal toBigDecimal(byte[] data, boolean desc) {
     }
 
     public static BigDecimal toBigDecimal(byte[] data, int dataOffset, int dataLength, boolean desc) {
-        byte[] dataCopy = Arrays.copyOfRange(data, dataOffset, dataOffset + dataLength);
-        if (desc) {
-            for (int i = 0; i < dataCopy.length; i++) {
-                dataCopy[i] ^= ORDER_FLIP_MASK;
-            }
-        }
-        return toBigDecimal(dataCopy, 0, dataLength - 1);
-    }
-
-    private static BigDecimal toBigDecimal(byte[] data, int dataOffset, int dataLength) {
-        int h = data[dataOffset];
-        h &= 0xff;
-        h >>>= Byte.SIZE - HEADER_SIZE_BITS;
+        int mask = desc ? ORDER_FLIP_MASK : 0;
+        int h = ((data[dataOffset] ^ mask) & 0xff) >>> (Byte.SIZE - HEADER_SIZE_BITS);
         if (h == HEADER_NULL) return null;
         if (h == HEADER_ZERO) return BigDecimal.ZERO;
 
-        byte sign = (byte) (h == HEADER_NEG ? -1 : 0);
+        dataLength -= 1;    // skip trailing 1
 
-        if (sign < 0) {
-            for (int z = 0; z < data.length; z++) {
-                if (z == 0) {
-                    data[z] ^= ORDER_FLIP_EXCLUDE_HEADER_MASK;
-                } else {
-                    data[z] ^= ORDER_FLIP_MASK;
-                }
-            }
-        }
+        boolean negative = (h == HEADER_NEG);
+        mask = desc ^ negative ? ORDER_FLIP_MASK : 0;
 
-        long[] expOffset = ScalarEncoding.readLong(data,dataOffset,HEADER_SIZE_BITS);
+        long[] expOffset = ScalarEncoding.readLong(data, dataOffset, desc ^ negative, HEADER_SIZE_BITS);
```

Comment:
Is this logic consistent with the original? If desc=false, and negative=true, the first two bytes will remain unchanged according to the original logic:
> if (z == 0) {	
>                     data[z] ^= ORDER_FLIP_EXCLUDE_HEADER_MASK;

But with the new logic, it will be flipped. Am I right?

Reply:
Yes, you are. Those two bits (not bytes) have already been processed, though, and are not looked at any more so there is no harm in changing them. The code, however, becomes simpler.

Reply:
Thanks @OlegMazurov , you are right that these two bits are not looked at anymore.